### PR TITLE
[Connector API] Support soft deletes of connectors

### PR DESCRIPTION
**Pull request by jedrazb**
### Add support for soft-deletes of connectors. Why?
- In order to improve UX on connectors on agentless we need to track what connectors were deleted to detect and propagate state change to fleet to add/remove corresponding policies

#### Changes
- define new system index `.connectors-deleted` that stores deleted connectors
- adapt `delete`, `get` and `list` operations logic to support this feature
- this change is backward compatible, nothing will be broken for existing users
- permission-wise it's also backward compatible, we still rely on index-level access to `.elastic-connectors-*` pattern 
  - Although this is likely to change soon once we actually enforce `manage_connector` and `monitor_connector` privileges 
- update docs
- add yaml tests and unit tests

#### Why new system index and not new `is_deleted` flag in existing index

The existing connector index is not a system index. Instead, we rely on index templates with non-dynamic mappings. This means that even if we update the template and incorporate logic to use the is_deleted field in the current index, it won’t be backward compatible, since the new field wouldn’t be present in mappings for users who created their connector indexes in the past. The is_deleted field will be included in the source, but we wouldn’t be able to run any queries against it, making it useless.

By using a separate index, we maintain backward compatibility for users who depend on direct access to the connector index. Deleted connectors will actually disappear from the original connector index and be moved into the deleted connectors index.

#### Followup work:
- update specification for 9.0 CLI

#### Other work we might need to do in the future:
- support force deletes, e.g pass `force=true` to delete endpoint to completely remove the connector

### Validation
- tested locally
- yaml e2e tests
- unit tests

### CODE REVIEW

File: muted-tests.yml [hunk 1]
```diff
@@ -97,8 +97,6 @@ tests:
 - class: org.elasticsearch.search.StressSearchServiceReaperIT
   method: testStressReaper
   issue: https://github.com/elastic/elasticsearch/issues/115816
-- class: org.elasticsearch.xpack.application.connector.ConnectorIndexServiceTests
-  issue: https://github.com/elastic/elasticsearch/issues/116087
```
Comment (artem-shelkovnikov) [inline muted-tests.yml:101]: What about unmuting the tests?

File: rest-api-spec/src/main/resources/rest-api-spec/api/connector.list.json [hunk 2]
```diff
@@ -47,6 +47,11 @@
       "query": {
         "type": "string",
         "description": "A search string for querying connectors, filtering results by matching against connector names, descriptions, and index names"
+      },
+      "deleted": {
+        "type": "boolean",
+        "default": false,
+        "description": "A flag indicating whether to list connectors that have been soft-deleted."
```
Comment (artem-shelkovnikov) [inline rest-api-spec/src/main/resources/rest-api-spec/api/connector.list.json:54]: Should we say that it's listing "only" deleted connectors, not all together?

File: x-pack/plugin/core/template-resources/src/main/resources/connectors-deleted.json [hunk 3]
```diff
@@ -0,0 +1,307 @@
+{
```
Comment (artem-shelkovnikov) [inline x-pack/plugin/core/template-resources/src/main/resources/connectors-deleted.json:1]: Is it possible to have a test that checks that both deleted and non-deleted indices have the same schema?

Additionally, does it make sense to make a lot of properties here searchable? Or we wanna have filters on deleted connectors too?

File: x-pack/plugin/ent-search/qa/rest/roles.yml [hunk 4]
```diff
@@ -2,6 +2,7 @@ admin:
   cluster:
     - manage_search_application
     - manage_behavioral_analytics
+    - manage_connector
```
Comment (artem-shelkovnikov) [inline x-pack/plugin/ent-search/qa/rest/roles.yml:5]: Can you add a comment on why this and below permissions are needed? What does it practically mean for users, if it does anything?

File: x-pack/plugin/ent-search/qa/rest/src/yamlRestTest/resources/rest-api-spec/test/entsearch/connector/20_connector_list.yml [hunk 5]
```diff
@@ -280,3 +280,137 @@ setup:
...
+      reason: Soft deletes were introduced in 9.0 release
+
+  - do:
+      connector.delete:
+        connector_id: connector-a
+
+  - do:
+      connector.list: {}
+
+  - match: { count: 2 }
```
Comment (artem-shelkovnikov) [inline x-pack/plugin/ent-search/qa/rest/src/yamlRestTest/resources/rest-api-spec/test/entsearch/connector/20_connector_list.yml:309]: Curious - why 2? It's some sort of initial setup?

File: x-pack/plugin/ent-search/qa/rest/src/yamlRestTest/resources/rest-api-spec/test/entsearch/connector/20_connector_list.yml [hunk 6]
```diff
@@ -280,3 +280,137 @@ setup:
...
+      connector.delete:
+        connector_id: connector-c
+
+  - do:
+      connector.list:
+        deleted: true
+
+  - match: { count: 3 }
+
+  # Alphabetical order by index_name for results
```
Comment (artem-shelkovnikov) [inline x-pack/plugin/ent-search/qa/rest/src/yamlRestTest/resources/rest-api-spec/test/entsearch/connector/20_connector_list.yml:339]: Curious - is it possible to do same matching without relying on sorting being stable?

File: x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/application/connector/ConnectorIndexService.java [hunk 7]
```diff
@@ -194,13 +242,19 @@ private Connector createConnectorWithDefaultValues(
      * Gets the {@link Connector} from the underlying index.
      *
      * @param connectorId The id of the connector object.
+     * @param isDeleted   If set to true, it returns only soft-deleted connector; otherwise, it returns non-deleted connector.
      * @param listener    The action listener to invoke on response/failure.
      */
-    public void getConnector(String connectorId, ActionListener<ConnectorSearchResult> listener) {
+    public void getConnector(String connectorId, Boolean isDeleted, ActionListener<ConnectorSearchResult> listener) {
```
Comment (artem-shelkovnikov) [inline x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/application/connector/ConnectorIndexService.java:248]: One interesting side effect of it is that it's possible to have a non-deleted and deleted connectors with same id that will actually point to really different connectors. That can cause some confusion, but probably nothing flow-breaking, other than customer restoring deleted connector manually breaking something, but that's on them? Should we document this more?

Comment (artem-shelkovnikov) [reply x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/application/connector/ConnectorIndexService.java:248]: Also in same scenario, deleting a live connector would erase previous record of a deleted connector. I guess it's also fine given our limitations, but might be nice to document.

**Review by artem-shelkovnikov — COMMENTED** [review]
I don't feel I can review Java code in the PR well, but I left a couple comments/questions about features introduced by the PR and other minor stuff

File: muted-tests.yml (hunk 1 above)
Comment (jedrazb) [reply muted-tests.yml:101]: It got muted due couple of flakey runs in the past ... I checked the logs and it seemed to be weird ES error. It has been running for almost a year now fine

**Review by kderusso — COMMENTED** [review]
I found this to be a little confusing. As a user I would assume that passing a deleted flag into a list call would have included all of the connectors including the soft deleted ones - ideally with metadata indicating which ones were deleted. Or, I would have expect a different API call to get only deletions. Why did we choose to combine these in an exclusive way?

Another option we could have taken is to add a deleted flag to the connectors index. Was this determined to be unmanageable?

**Comment by jedrazb** [comment]
We’ve discussed this with Jim, and there might be a better way to implement logic to handle soft-deleted documents, with mappings update. I’m implementing his suggestion as a separate branch of this PR. Suggested approach involves updating the template and applying new mappings if needed using the onClusterChange listener.

**Comment by jedrazb** [comment]
@kderusso Thank you for looking into this!

> Another option we could have taken is to add a deleted flag to the connectors index. Was this determined to be unmanageable?

Jim made me aware that there is a way to upgrade connector index template mappings in clusterChanged logic, so we can just add a new field to the existing index. I’m following this approach and will have something to share soon.

> As a user I would assume that passing a deleted flag into a list call would have included all of the connectors including the soft deleted ones - ideally with metadata indicating which ones were deleted

I don’t have a strong opinion here. For me, passing `deleted=true` could be interpreted as either:
1.	“Show me (filter)” only the deleted ones, or
2.	Include the deleted ones in the response with a flag indicating they were deleted.

Do you know if ES endpoints conventions prefer either approach?

**Comment by jedrazb** [comment]
Closing in favour of https://github.com/elastic/elasticsearch/pull/118669

**Pull request closed without merging**